src/Text/Parsing/Parser/Combinators.purs

```diff
@@ -168,51 +184,129 @@ sepEndBy1Rec p sep = do
 endBy1 :: forall m s a sep. Monad m => ParserT s m a -> ParserT s m sep -> ParserT s m (NonEmptyList a)
 endBy1 p sep = many1 $ p <* sep
 
+-- | Stack-safe version of `endBy1` at the expense of `MonadRec` constraint
+endBy1Rec :: forall m s a sep. MonadRec m => ParserT s m a -> ParserT s m sep -> ParserT s m (NonEmptyList a)
+endBy1Rec p sep = many1Rec $ p <* sep
+
 -- | Parse phrases delimited and terminated by a separator.
 endBy :: forall m s a sep. Monad m => ParserT s m a -> ParserT s m sep -> ParserT s m (List a)
 endBy p sep = many $ p <* sep
 
+-- | Stack-safe version of `endBy` at the expense of `MonadRec` constraint
+endByRec :: forall m s a sep. MonadRec m => ParserT s m a -> ParserT s m sep -> ParserT s m (List a)
+endByRec p sep = manyRec $ p <* sep
+
 -- | Parse phrases delimited by a right-associative operator.
 -- |
 -- | For example:
 -- |
 -- | ```purescript
--- | chainr digit (string "+" *> add) 0
+-- | chainr digit (string "+" $> add) 0
 -- | ```
 chainr :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
 chainr p f a = chainr1 p f <|> pure a
 
+-- | Stack-safe version of `chainr` at the expense of `MonadRec` constraint.
+chainrRec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
+chainrRec p f a = chainr1Rec p f <|> pure a
+
 -- | Parse phrases delimited by a left-associative operator.
+-- |
+-- | For example:
+-- |
+-- | ```purescript
+-- | chainr digit (string "+" $> add) 0
+-- | ```
 chainl :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
 chainl p f a = chainl1 p f <|> pure a
 
+-- | Stack-safe version of `chainl` at the expense of `MonadRec` constraint.
+chainlRec :: forall m s a. MonadRec m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
+chainlRec p f a = chainl1Rec p f <|> pure a
+
 -- | Parse phrases delimited by a left-associative operator, requiring at least one match.
 chainl1 :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> ParserT s m a
 chainl1 p f = do
   a <- p
-  chainl1' p f a
-
-chainl1' :: forall m s a. Monad m => ParserT s m a -> ParserT s m (a -> a -> a) -> a -> ParserT s m a
-chainl1' p f a =
-  ( do
-      f' <- f
-      a' <- p
-      chainl1' p f (f' a a')
-  ) <|> pure a
+  chainl1' a
+  where
+  chainl1' a =
+    ( do
+        f' <- f
+        a' <- p
+        chainl1' (f' a a')
+    ) <|> pure a
```

Comment:
I moved this under `where`, because it seems like a private helper function. Technically it might be breaking though, if anybody happens to use this function for some reason. Should it be added to changelog? Or would you prefer it be brought back to the top level?

Reply:
Yeah, technically it is breaking, isn't it.

I'll add this to the changelog and we'll bump major version to 8.

Reply:
It looks like this private helper function has been exported since 2014, when this file was renamed from something else.

I agree that this is a private helper function and it should be private.

Reply:
It's better this way, thanks.